sound/soc/sof/intel/hda-loader.c

```diff
@@ -61,11 +62,14 @@ hda_cl_prepare(struct device *dev, unsigned int format, unsigned int size,
 	hstream = &hext_stream->hstream;
 	hstream->substream = NULL;
 
-	/* allocate DMA buffer */
-	ret = snd_dma_alloc_pages(SNDRV_DMA_TYPE_DEV_SG, dev, size, dmab);
-	if (ret < 0) {
-		dev_err(sdev->dev, "error: memory alloc failed: %d\n", ret);
-		goto out_put;
+	if (!keep_buffer || (keep_buffer && !dmab->area)) {
```

Comment:
this sort of test gives me a migraine.

Reply:
We need to allocate the DMA buffer if we don't want to keep the DMA buffer OR if we want to keep the buffer, but it has not yet been allocated.
I can add a comment.